keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+
+### Non Goals
+
+
+1. This KEP focuses on the core resource types of CPU and memory. It doesn’t intend to address other resource types (e.g. GPU, storage, network bandwidth)
+at pod level in this phase. However, it could be considered in future extensions of the KEP.
+
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+A development environment implemented as a Kubernetes Pod allows for separation of tools and a (web-based) IDE between multiple side-cars.
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory.
+Additional tools are provided in additional side-car deployments - for example an LSP( https://langserver.org/) service, a terminal, and more. 
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit
+defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024Mi"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+
+#### Story 2
+
+Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
+level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
+not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
+need to over-provision for each container’s peak usage. 
+
+
+### Notes/Constraints/Caveats
+
+
+1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
+2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
+turn it on by specifying the relevant resource section at pod-level in the Pod specification.
+
+
+### Risks and Mitigations
+
+Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
```

Comment:
From scheduling and queueing perspective, this will impact all kinds of schedulers and queueing systems(kueue etc.), so this is a potential risk I think or something we should highlight.

The default scheduler depends on the `PodRequests` right now, https://github.com/kubernetes/kubernetes/blob/a81ea5d0af41a7e519a06780aced3c517f61547c/pkg/api/v1/resource/helpers.go#L50, maybe we should expose the function for widely usage to avoid the redundant work.

cc @alculquicondor

Reply:
Good callout. @alculquicondor could you help review the KEP for sig-scheduling? 

For alpha, it should be okay even if other schedulers/queuing systems don't support the feature. But for beta/GA, it'd be good if we can work out a plan for others to easily support this.

Reply:
The most related part is the nodeResource plugin, for null-request Pods, we will apply with a default value for each container to avoid scheduling too many null-request Pods because they're consuming resources actually. This will be conflict with the Pod level requests.

We should change the logic here.

Reply:
Previous discussions around homogenizing the calculation function are in the PR https://github.com/kubernetes/kubernetes/pull/115367

Reply:
ah, there was actually this open PR https://github.com/kubernetes/kubernetes/pull/124609

Reply:
More defaulting logic that we need to take into account https://github.com/kubernetes/kubernetes/blob/f248c24456254d6e0668a424a31258e6d9e9c185/pkg/scheduler/framework/types.go#L927